python/ray/train/_internal/state/state_manager.py

```diff
@@ -82,14 +85,22 @@ def collect_train_worker_info():
             for ds_name, ds in datasets.items()
         ]
 
-        train_run_info = TrainRunInfo(
+        self.train_run_info_dict = dict(
             id=run_id,
             job_id=job_id,
             name=run_name,
             controller_actor_id=controller_actor_id,
             workers=worker_info_list,
             datasets=dataset_info_list,
             start_time_ms=start_time_ms,
+            run_status=run_status,
+            status_detail=status_detail,
         )
+        train_run_info = TrainRunInfo(**self.train_run_info_dict)
+        ray.get(self.state_actor.register_train_run.remote(train_run_info))
 
+    def update_train_run_info(self, updates: Dict[str, Any]) -> None:
+        """Update specific fields of a registered TrainRunInfo instance."""
+        self.train_run_info_dict.update(updates)
+        train_run_info = TrainRunInfo(**self.train_run_info_dict)
         ray.get(self.state_actor.register_train_run.remote(train_run_info))
```

Comment:
This can cause problems if not careful in the future e.g. if somewhere we call `update_train_run_info` with a different `id`. One way to limit this is to restrict the public API and use a private method instead.

Something like this:
```suggestion
        updates = dict(
            id=run_id,
            job_id=job_id,
            name=run_name,
            controller_actor_id=controller_actor_id,
            workers=worker_info_list,
            datasets=dataset_info_list,
            start_time_ms=start_time_ms,
            run_status=run_status,
            status_detail=status_detail,
        )
        # you can first assert that the info dict is empty here I guess
        _update_train_run_info(updates)

    def end_train_run(self, run_status, status_detail, end_time_ms):
        updates=dict(
            run_status=run_status,
            status_detail=status_detail,
            end_time_ms=end_time_ms,
        )
        _update_train_run_info(updates)

    def _update_train_run_info(self, updates: Dict[str, Any]) -> None:
        """Update specific fields of a registered TrainRunInfo instance."""
        self.train_run_info_dict.update(updates)
        train_run_info = TrainRunInfo(**self.train_run_info_dict)
        ray.get(self.state_actor.register_train_run.remote(train_run_info))
```

Reply:
This is a good point! Let me update this